File: src/core/memory.rs
use crate::core::timer::Timer;

pub struct Memory {
    memory: [u8; 0xFFFF],
    bootrom: [u8; 0x100],
    IE: u8,
    IF: u8,
    pub bootrom_loaded: bool
}

impl Memory {
    pub fn new() -> Self {
        Self {
            memory: [0; 0xFFFF],
            bootrom: [0; 0x100],
            bootrom_loaded: false,
            IE: 0x00,
            IF: 0x00
        }
    }

    pub fn write(&mut self, location: u16, value: u8, timer: &mut Timer) {
        let location = location as usize;
        if location >= self.memory.len() { }
        else if location == 0xFF02 && value == 0x81  {
            print!("{}", self.memory[0xFF01] as char);
        }
        else if location >= 0xFF04 && location <= 0xFF07  {

        } 
        else if location == 0xFFFF {
            self.IE = value;
        }
        else if location == 0xFF0F {
            self.IF = value;
        }
        else if location >= 0xFF04 && location <= 0xFF07 {
            timer.write(location, value)
        }
        else {
            self.memory[location] = value
        }
    }

    pub fn read(&self, location: u16, timer: &mut Timer) -> u8 {
        let location: usize = location as usize;
        if location >= self.memory.len() {
            return 0;
        }

        if location < 0x100 && self.bootrom_loaded {
            return self.bootrom[location];
        }

        if location == 0xFFFF {
            return self.IE;
        }

        if location == 0xFF0F {
            return self.IF;
        }

        if location >= 0xFF04 && location <= 0xFF07 {
            let value = timer.read(location);
            return value;
        }

        self.memory[location]
    }

    pub fn write_bootrom(&mut self, bootrom: &[u8; 0x100]) {
        let mut counter = 0;

        for x in bootrom {
            self.bootrom[counter] = *x;
            counter += 1;
        }
        self.bootrom_loaded = true;
    }

    pub fn write_game(&mut self, game: &[u8]) {
        let mut counter = 0;

        for x in game {
            self.memory[counter] = *x;
            counter += 1;
        }
    }
}

File: Sources/RetroDMG/Memory.swift
//
//  Memory.swift
//
//
//  Created by Glenn Hevey on 24/12/2023.
//

import Foundation

struct Memory {
    var memory: [UInt8]
    var bootrom: [UInt8]
    var interruptEnabled: UInt8
    var interruptFlag: UInt8
    var bootromLoaded: Bool
    
    init() {
        memory = [UInt8](repeating: 0, count: 65537)
        bootrom = [UInt8](repeating: 0, count: 0x100)
        bootromLoaded = false
        interruptEnabled = 0x00
        interruptFlag = 0x00
    }
    
    mutating func write(location: UInt16, value: UInt8) {
        if location >= memory.endIndex {
            print("\(location.hex) is out of bounds")
            return
        }
        
        if location == 0xFF02 && value == 0x81 {
            print(memory[0xFF01])
        } else {
            memory[Int(location)] = value
        }
    }
    
    mutating func write(bootrom: [UInt8]) {
        self.bootrom = bootrom
    }
    
    mutating func write(rom: [UInt8]) {
        memory.replaceSubrange(0...rom.count, with: rom)
    }
    
    func read(location: UInt16) -> UInt8 {
        let location = Int(location)
        if location > memory.count {
            return 0
        }
        
        if location < 0x100 && bootromLoaded {
           return bootrom[location]
        }
        
        return memory[location]
    }
}
